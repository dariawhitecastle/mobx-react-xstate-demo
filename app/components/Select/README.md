<!-- ! This is a generated file. To make changes, edit <Component>.doc.js ! -->
## CreateSelect
A control to select a value, with optional search.

## Usage

```javascript
import { Select } from 'grommet';
<Select />
```

## Properties

**a11yTitle**

Custom title to be used by screen readers.

```
string
```

**alignSelf**

How to align along the cross axis when contained in
      a Box or along the column axis when contained in a Grid.

```
start
center
end
stretch
```

**gridArea**

The name of the area to place
    this inside a parent Grid.

```
string
```

**margin**

The amount of margin around the component. An object can
      be specified to distinguish horizontal margin, vertical margin, and
      margin on a particular side.

```
none
xxsmall
xsmall
small
medium
large
xlarge
{
  bottom: 
    xxsmall
    xsmall
    small
    medium
    large
    xlarge
    string,
  horizontal: 
    xxsmall
    xsmall
    small
    medium
    large
    xlarge
    string,
  left: 
    xxsmall
    xsmall
    small
    medium
    large
    xlarge
    string,
  right: 
    xxsmall
    xsmall
    small
    medium
    large
    xlarge
    string,
  top: 
    xxsmall
    xsmall
    small
    medium
    large
    xlarge
    string,
  vertical: 
    xxsmall
    xsmall
    small
    medium
    large
    xlarge
    string
}
string
```

**children**

Function that will be called when each option is rendered.
      It will be passed (option, index, options, state) where option
      is the option to render, index is the index of that option in the
      options array, and state is an object with
      { active, disabled, selected } keys indicating the current state
      of the option.

```
function
```

**closeOnChange**

Wether to close the drop when a selection is made. Defaults to `true`.

```
boolean
```

**disabled**

Whether the entire select or individual options should be disabled.
        An array of numbers indicates the indexes into 'options' of the
        disabled options. An array of strings or objects work the same way
        as the 'value' to indicate which options are disabled.

```
boolean
[
  number
  string
  object
]
```

**disabledKey**

When the options array contains objects, this property indicates how
      to determine which options should be disabled. If a string is
      provided, it is used as the key for each item object and if that key
      returns truthy, the option is disabled. If a function is provided, it is
      called with the option and the return value determines if the option
      is disabled.

```
string
function
```

**dropAlign**

How to align the drop. Defaults to `{
  "top": "bottom",
  "left": "left"
}`.

```
{
  top: 
    top
    bottom,
  bottom: 
    top
    bottom,
  right: 
    left
    right,
  left: 
    left
    right
}
```

**dropHeight**

The height of the drop container.

```
xsmall
small
medium
large
xlarge
string
```

**dropTarget**

Target where the options drop will be aligned to. This should be
      a React reference. Typically, this is not required as the drop will be
      aligned to the Select itself by default.

```
object
```

**dropProps**

Any valid Drop prop.

```
object
```

**focusIndicator**

Whether when 'plain' it should receive a focus outline.

```
boolean
```

**icon**

A custom icon to be used when rendering the select. You can use false to not render an icon at all.

```
boolean
function
node
```

**labelKey**

When the options array contains objects, this property indicates how
      to determine the label of each option. If a string is
      provided, it is used as the key to retrieve each option's label.
      If a function is provided, it is called with the option and the
      return value indicates the label.

```
string
function
```

**messages**

Custom messages.

```
{
  multiple: string
}
```

**multiple**

Whether to allow multiple options to be selected.

```
boolean
```

**onChange**

Function that will be called when the user selects an option.

```
function
```

**onClose**

Function that will be called when the Select drop closes.

```
function
```

**onOpen**

Function that will be called when the Select drop opens.

```
function
```

**onSearch**

Function that will be called when the user types in the search input.
      If this property is not provided, no search field will be rendered.

```
function
```

**onMore**

Use this to indicate that 'items' doesn't contain all that it could.
      It will be called when the entire list of items has been rendered.
      This might be used when the total number of items that could be retrieved
      is more than you'd want to load into the browser. 'onMore' allows you
      to lazily fetch more from the server only when needed.

```
function
```

**replace**

Whether to replace previously rendered items with a generic spacing
      element when they have scrolled out of view. This is more performant but
      means that in-page searching will not find elements that have been
      replaced. Defaults to `true`.

```
boolean
```

**options**

Required. Options can be either a string or an object. If an object is used, use
      children callback in order to render anything based on the current item.

```
[
  string
  element
  object
]
```

**open**

Control the state of the component.

```
boolean
```

**placeholder**

Placeholder text to use when no value is provided.

```
string
node
element
```

**plain**

Whether this is a plain Select input with no border or padding.

```
boolean
```

**searchPlaceholder**

Placeholder text to use in the search box when the search input is empty.

```
string
```

**selected**

Index of the currently selected option. When multiple, the set of
      options selected. NOTE: This is deprecated in favor of indicating
      the selected values via the 'value' property.

```
number
[number]
```

**size**

The size of the select.

```
small
medium
large
xlarge
string
```

**value**

Currently selected value. This can be an array
      when multiple. Passing an element allows the caller to control how
      the value is rendered. Passing an element is deprecated. Instead,
      use the 'valueLabel' property.

```
string
element
object
[
  string
  object
]
```

**valueLabel**

Provides custom rendering of the value. If not provided, Select
      will render the value automatically.

```
node
```

**valueKey**

When the options array contains objects, this property indicates how
      to determine the value of each option. If a string is
      provided, it is used as the key to retrieve each option's value.
      If a function is provided, it is called with the option and the
      return value indicates the value.

```
string
function
```

**emptySearchMessage**

Empty option message to display when no matching results were found Defaults to `No matches found`.

```
string
```
  
## Theme
  
**select.background**

The background color used for Select. Expects `string`.

Defaults to

```
undefined
```

**select.options.container**

Any valid Box prop for the options container. Expects `object`.

Defaults to

```
{ align: 'start', pad: 'small' }
```

**select.options.text**

Any valid Text prop for text used inside the options container. Expects `object`.

Defaults to

```
{ margin: 'none }
```

**select.container.extend**

Any additional style for the container of the Select component. Expects `string | (props) => {}`.

Defaults to

```
undefined
```

**select.control.extend**

Any additional style for the control of the Select component. Expects `string | (props) => {}`.

Defaults to

```
undefined
```

**select.icons.margin**

The margin used for Select icons. Expects `string | object`.

Defaults to

```
undefined
```

**select.icons.color**

The color used for Select icons. Expects `string | { dark: string, light: string }`.

Defaults to

```
undefined
```

**select.icons.down**

The down icon to use for opening the Select. Expects `React.Element`.

Defaults to

```
<FormDown />
```

**select.searchInput**

Component for the Select search input field. Expects `React.Component`.

Defaults to

```
undefined
```

**select.step**

How many items to render at a time. Expects `number`.

Defaults to

```
20
```
