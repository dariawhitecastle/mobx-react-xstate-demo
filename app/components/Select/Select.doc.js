import { describe, PropTypes } from 'react-desc'

import { genericProps } from '../../utils/propTypes'

export const doc = Select => {
  const DocumentedSelect = describe(Select)
    .description('A control to select a value, with optional search.')
    .usage(
      `import { Select } from 'grommet';
<Select />`,
    )
  // We don't include svg due to a collision on the values property
  // .intrinsicElement('select');

  DocumentedSelect.propTypes = {
    ...genericProps,
    children: PropTypes.func.description(
      `Function that will be called when each option is rendered.
      It will be passed (option, index, options, state) where option
      is the option to render, index is the index of that option in the
      options array, and state is an object with
      { active, disabled, selected } keys indicating the current state
      of the option.`,
    ),
    closeOnChange: PropTypes.bool
      .description('Wether to close the drop when a selection is made.')
      .defaultValue(true),
    disabled: PropTypes.oneOfType([
      PropTypes.bool,
      PropTypes.arrayOf(
        PropTypes.oneOfType([PropTypes.number, PropTypes.string, PropTypes.object]),
      ),
    ])
      .description(
        `Whether the entire select or individual options should be disabled.
        An array of numbers indicates the indexes into 'options' of the
        disabled options. An array of strings or objects work the same way
        as the 'value' to indicate which options are disabled.`,
      )
      .defaultValue(false),
    disabledKey: PropTypes.oneOfType([PropTypes.string, PropTypes.func]).description(
      `When the options array contains objects, this property indicates how
      to determine which options should be disabled. If a string is
      provided, it is used as the key for each item object and if that key
      returns truthy, the option is disabled. If a function is provided, it is
      called with the option and the return value determines if the option
      is disabled.`,
    ),
    dropAlign: PropTypes.shape({
      top: PropTypes.oneOf(['top', 'bottom']),
      bottom: PropTypes.oneOf(['top', 'bottom']),
      right: PropTypes.oneOf(['left', 'right']),
      left: PropTypes.oneOf(['left', 'right']),
    })
      .description('How to align the drop.')
      .defaultValue({
        top: 'bottom',
        left: 'left',
      }),
    dropHeight: PropTypes.oneOfType([
      PropTypes.oneOf(['xsmall', 'small', 'medium', 'large', 'xlarge']),
      PropTypes.string,
    ]).description('The height of the drop container.'),
    dropTarget: PropTypes.object.description(
      `Target where the options drop will be aligned to. This should be
      a React reference. Typically, this is not required as the drop will be
      aligned to the Select itself by default.`,
    ),
    dropProps: PropTypes.object.description('Any valid Drop prop.'),
    focusIndicator: PropTypes.bool.description(
      "Whether when 'plain' it should receive a focus outline.",
    ),
    icon: PropTypes.oneOfType([PropTypes.bool, PropTypes.func, PropTypes.node]).description(
      'A custom icon to be used when rendering the select. You can use false to not render an icon at all.',
    ),
    labelKey: PropTypes.oneOfType([PropTypes.string, PropTypes.func]).description(
      `When the options array contains objects, this property indicates how
      to determine the label of each option. If a string is
      provided, it is used as the key to retrieve each option's label.
      If a function is provided, it is called with the option and the
      return value indicates the label.`,
    ),
    messages: PropTypes.shape({
      multiple: PropTypes.string,
    }).description('Custom messages.'),
    multiple: PropTypes.bool.description('Whether to allow multiple options to be selected.'),
    onChange: PropTypes.func.description(
      'Function that will be called when the user selects an option.',
    ),
    onClose: PropTypes.func.description(
      'Function that will be called when the Select drop closes.',
    ),
    onOpen: PropTypes.func.description('Function that will be called when the Select drop opens.'),
    onSearch: PropTypes.func.description(
      `Function that will be called when the user types in the search input.
      If this property is not provided, no search field will be rendered.`,
    ),
    onMore: PropTypes.func.description(
      `Use this to indicate that 'items' doesn't contain all that it could.
      It will be called when the entire list of items has been rendered.
      This might be used when the total number of items that could be retrieved
      is more than you'd want to load into the browser. 'onMore' allows you
      to lazily fetch more from the server only when needed.`,
    ),
    replace: PropTypes.bool
      .description(
        `Whether to replace previously rendered items with a generic spacing
      element when they have scrolled out of view. This is more performant but
      means that in-page searching will not find elements that have been
      replaced.`,
      )
      .defaultValue(true),
    options: PropTypes.arrayOf(
      PropTypes.oneOfType([PropTypes.string, PropTypes.element, PropTypes.object]),
    ).description(
      `Options can be either a string or an object. If an object is used, use
      children callback in order to render anything based on the current item.`,
    ).isRequired,
    open: PropTypes.bool.description(`Control the state of the component.`),
    placeholder: PropTypes.oneOfType([
      PropTypes.string,
      PropTypes.node,
      PropTypes.element,
    ]).description('Placeholder text to use when no value is provided.'),
    plain: PropTypes.bool.description(
      'Whether this is a plain Select input with no border or padding.',
    ),
    searchPlaceholder: PropTypes.string.description(
      'Placeholder text to use in the search box when the search input is empty.',
    ),
    selected: PropTypes.oneOfType([
      PropTypes.number,
      PropTypes.arrayOf(PropTypes.number),
    ]).description(
      `Index of the currently selected option. When multiple, the set of
      options selected. NOTE: This is deprecated in favor of indicating
      the selected values via the 'value' property.`,
    ),
    size: PropTypes.oneOfType([
      PropTypes.oneOf(['small', 'medium', 'large', 'xlarge']),
      PropTypes.string,
    ]).description('The size of the select.'),
    value: PropTypes.oneOfType([
      PropTypes.string,
      PropTypes.element, // deprecated, use valueLabel
      PropTypes.object,
      PropTypes.arrayOf(PropTypes.oneOfType([PropTypes.string, PropTypes.object])),
    ]).description(`Currently selected value. This can be an array
      when multiple. Passing an element allows the caller to control how
      the value is rendered. Passing an element is deprecated. Instead,
      use the 'valueLabel' property.`),
    valueLabel: PropTypes.node.description(
      `Provides custom rendering of the value. If not provided, Select
      will render the value automatically.`,
    ),
    valueKey: PropTypes.oneOfType([PropTypes.string, PropTypes.func]).description(
      `When the options array contains objects, this property indicates how
      to determine the value of each option. If a string is
      provided, it is used as the key to retrieve each option's value.
      If a function is provided, it is called with the option and the
      return value indicates the value.`,
    ),
    emptySearchMessage: PropTypes.string
      .description(`Empty option message to display when no matching results were found`)
      .defaultValue('No matches found'),
  }

  return DocumentedSelect
}

export const themeDoc = {
  'select.background': {
    description: 'The background color used for Select.',
    type: 'string',
    defaultValue: undefined,
  },
  'select.options.container': {
    description: 'Any valid Box prop for the options container.',
    type: 'object',
    defaultValue: "{ align: 'start', pad: 'small' }",
  },
  'select.options.text': {
    description: 'Any valid Text prop for text used inside the options container.',
    type: 'object',
    defaultValue: "{ margin: 'none }",
  },
  'select.container.extend': {
    description: 'Any additional style for the container of the Select component.',
    type: 'string | (props) => {}',
    defaultValue: undefined,
  },
  'select.control.extend': {
    description: 'Any additional style for the control of the Select component.',
    type: 'string | (props) => {}',
    defaultValue: undefined,
  },
  'select.icons.margin': {
    description: 'The margin used for Select icons.',
    type: 'string | object',
    defaultValue: undefined,
  },
  'select.icons.color': {
    description: 'The color used for Select icons.',
    type: 'string | { dark: string, light: string }',
    defaultValue: undefined,
  },
  'select.icons.down': {
    description: 'The down icon to use for opening the Select.',
    type: 'React.Element',
    defaultValue: '<FormDown />',
  },
  'select.searchInput': {
    description: `Component for the Select search input field.`,
    type: 'React.Component',
    defaultValue: undefined,
  },
  'select.step': {
    description: `How many items to render at a time.`,
    type: 'number',
    defaultValue: 20,
  },
}
