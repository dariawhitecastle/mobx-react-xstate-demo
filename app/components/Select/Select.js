/* eslint-disable react/prop-types */
/**
 *
 * Select
 *
 */

import React, { useState } from 'react'
import { Select } from 'grommet'
import { doc } from './Select.doc'
import helpers from '../../utils/helpers'

const CreateSelect = ({
  a11yTitle,
  alignSelf,
  children,
  closeOnChange,
  disabled,
  dropAlign,
  dropProps,
  dropHeight,
  dropTarget,
  emptySearchMessage,
  focusIndicator,
  labelKey,
  icon,
  margin,
  messages,
  multiple,
  onChange,
  onClose,
  onMore,
  onOpen,
  onSearch,
  open,
  options,
  placeholder,
  plain,
  replace,
  searchPlaceholder,
  selected,
  size,
  value,
  ...rest
}) => {
  const [filteredOptions, setFilteredOptions] = useState(options)
  return (
    <Select
      a11yTitle={a11yTitle}
      alignSelf={alignSelf}
      closeOnChange={closeOnChange}
      disabled={disabled}
      dropAlign={dropAlign}
      dropProps={dropProps}
      dropHeight={dropHeight}
      dropTarget={dropTarget}
      emptySearchMessage={emptySearchMessage}
      focusIndicator={focusIndicator}
      labelKey={labelKey}
      icon={icon}
      margin={margin}
      messages={messages}
      multiple={multiple}
      onChange={onChange}
      onClose={onClose}
      onOpen={onOpen}
      onSearch={searchText => {
        const regexp = new RegExp(searchText, 'i')
        setFilteredOptions(options.filter(o => o.match(regexp)))
      }}
      open={open}
      options={filteredOptions}
      placeholder={placeholder}
      plain={plain}
      replace={replace}
      searchPlaceholder={searchPlaceholder}
      selected={selected}
      size={size}
      value={value}
      {...rest}
    >
      {children}
    </Select>
  )
}

export default helpers.createWithDoc({
  docFunction: doc,
  component: CreateSelect,
})
