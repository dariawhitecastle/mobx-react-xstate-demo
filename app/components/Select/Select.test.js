/**
 *
 * Tests for Select
 *
 * @see https://github.com/react-boilerplate/react-boilerplate/tree/master/docs/testing
 *
 */

import React from 'react'
import { render, fireEvent } from '@testing-library/react'
// import 'jest-dom/extend-expect'; // add some helpful assertions

import Select from './Select'

const options = ['one', 'two', 'three']
const renderComponent = (props = {}) =>
  render(<Select label="Select" {...props} options={options} />)

const setup = () => {
  const utils = render(<Select placeholder="select" options={options} />)
  const select = utils.getByPlaceholderText('select')
  return {
    select,
    ...utils,
  }
}

test('it renders and matches snapshot', () => {
  const { container } = renderComponent()
  expect(container).toMatchSnapshot()
})

test('It should update value on select', () => {
  const { select } = setup()
  expect(select.value).toBe('') // empty before
  fireEvent.select(select, { target: { value: 'one' } })
  expect(select.value).toBe('one')
})
