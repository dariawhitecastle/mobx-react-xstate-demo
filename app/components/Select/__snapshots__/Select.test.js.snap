// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`it renders and matches snapshot 1`] = `
<div>
  <button
    aria-label="Open Drop"
    class="Select__StyledSelectDropButton-sc-17idtfo-1 fcNjcL StyledButton-sc-323bzc-0 jfbKsv"
    type="button"
  >
    <div
      class="StyledBox-sc-13pk1d4-0 erazuK"
    >
      <div
        class="StyledBox-sc-13pk1d4-0 XINUX"
      >
        <div
          class="StyledTextInput__StyledTextInputContainer-sc-1x30a0s-1 kYQCpm"
        >
          <input
            autocomplete="off"
            class="Select__SelectTextInput-sc-17idtfo-0 lmyIUy StyledTextInput-sc-1x30a0s-0 duvNfq"
            label="Select"
            readonly=""
            tabindex="-1"
            type="text"
            value=""
          />
        </div>
      </div>
      <div
        class="StyledBox-sc-13pk1d4-0 eLQtUu"
        style="min-width: auto;"
      >
        <svg
          aria-label="FormDown"
          class="StyledIcon-ofa7kd-0 dSllCE"
          viewBox="0 0 24 24"
        >
          <polyline
            fill="none"
            points="18 9 12 15 6 9"
            stroke="#000"
            stroke-width="2"
          />
        </svg>
      </div>
    </div>
  </button>
</div>
`;
